### DISCUSSION TITLE
How to dynamically add pages or components?

### DISCUSSION THREAD

**Discussion by rgathmann** [QuestPDF/QuestPDF#34] [Q&A] [answered]
I love the library, but I am struggling to figure out the best way to dynamically add pages or components. For example, lets say you had Page A, Page B, and Page C and we wanted to mix and match these when creating the document. 

My first thought would be to just have the content for each page as a separate component and pass them in as a List<IComponent> when construction the document. Then during compose you would simple iterate over the list of components and create a page for each. However this fails. Is there a better way to do this?

```csharp
public class ModernPlannerDocument: IDocument
    {
        private List<IComponent> PageComponents;

        public ModernPlannerDocument()
        {
            PageComponents = new List<IComponent>();
        }

        public void Compose(IDocumentContainer container)
        {
            foreach(var page in PageComponents)
            {
                container.Page(page =>
                {
                    page.Content().Component(page);

                });
            }
        }

        public void AddPage(IComponent content)
        {

        }

        public DocumentMetadata GetMetadata() => DocumentMetadata.Default;
    }
```

**Comment by MarcinZiabek** [comment, marked as answer]
Hello, can you please provide more insights into your use case? If there is any kind of exception, please share :)

I suspect that you are encountering the variable closure conflict (with the variable `page`). Please try with this code:

```csharp
public void Compose(IDocumentContainer container)
{
    foreach(var component in PageComponents) // renamed 'page' to 'component'
    {
        container.Page(page =>
        {
            page.Content().Component(component);

        });
    }
}
```

**Reply by rgathmann** [reply]
This was the answer. Please feel free to delete this thread, it was a silly mistake that will probably not be helpful to others. 🤦

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
